File: 0027-remove-element/0027-remove-element.ts
function removeElement(nums: number[], val: number): number {
  let [ahead, behind] = [0, 0]
  while (ahead < nums.length) {
    while (ahead < nums.length && nums[ahead] === val) ahead++
    while (ahead < nums.length && nums[ahead] !== val) {
      nums[behind] = nums[ahead]
      behind++
      ahead++
    }
  }
  return behind
}

File: 0027-remove-element/0027-remove-element.go
func removeElement(nums []int, val int) int {
  behind, ahead := 0, 0
  for ahead < len(nums) {
    for ahead < len(nums) && nums[ahead] == val {
      ahead++
    }
    for ahead < len(nums) && nums[ahead] != val {
      nums[behind] = nums[ahead]
      behind++
      ahead++
    }
  }
  return behind
}